Rupture - Breach- Realtime service failing to process the html response from backend

We have the following bug report from Murali:

----

Hi Jenius/Dionyziz/Dimit,
    This is Murali, I attended the BlackHat Asia 2016 at Marina BaySands, Singapore. 
  BREACH is one the best topics of Black Hat Asia briefings.I appreciate for the great work done by you guys.

I ran into issues while trying to test the breach attack with provided client/test.html. I'm facing below issue in realtime services and realtime service is halting immediately first request serving.

I followed the ReadME.md instruction. 

Step 1: 
 gulp watchify
Step 2: (realtime service)
 npm start
Step 3:
  python manage.py runserver
Step 4:
 Lunch test.html in the browser.

 

I can see the request flowed from client--> realtime --> backend and realtime service is halting due to expected response type JSON data. 

It looks like realtime service is expecting the JSON data from backend but where as backend is sending html response.

I managed to run realtime service by handling the error at JSON response data handling logic. But not getting the expected result as show in youtube video or demo shared in briefing. 

How to do I ensure to make it work and get the expected result. 

I greatly appreciate your help on this.

Cheers,
Murali.

MacBook-Pro:realtime test$ npm start

> rupture@1.0.0 start /Users/test/M3/rupture-develop/realtime
> node --harmony --harmony_destructuring index.js

2016-05-06T09:54:55.111Z - info: Rupture real-time service starting
2016-05-06T09:54:55.113Z - info: Listening on port 3031
2016-05-06T09:55:59.036Z - info: New connection from client /#KzbLdRW25lq5vPcUAAAA
2016-05-06T09:55:59.074Z - info: get-work from client /#KzbLdRW25lq5vPcUAAAA
2016-05-06T09:55:59.193Z - info: Got (get-work) response from backend: 
```
<!DOCTYPE html>
<html lang="en">
<head>
  <meta http-equiv="content-type" content="text/html; charset=utf-8">
  <title>Page not found at /breach/get_work/1</title>
  <meta name="robots" content="NONE,NOARCHIVE">
  <style type="text/css">
    html * { padding:0; margin:0; }
    body * { padding:10px 20px; }
    body * * { padding:0; }
    body { font:small sans-serif; background:#eee; }
    body>div { border-bottom:1px solid #ddd; }
    h1 { font-weight:normal; margin-bottom:.4em; }
    h1 span { font-size:60%; color:#666; font-weight:normal; }
    table { border:none; border-collapse: collapse; width:100%; }
    td, th { vertical-align:top; padding:2px 3px; }
    th { width:12em; text-align:right; color:#666; padding-right:.5em; }
    #info { background:#f6f6f6; }
    #info ol { margin: 0.5em 4em; }
    #info ol li { font-family: monospace; }
    #summary { background: #ffc; }
    #explanation { background:#eee; border-bottom: 0px none; }
  </style>
</head>
<body>
  <div id="summary">
    <h1>Page not found <span>(404)</span></h1>
    <table class="meta">
      <tr>
        <th>Request Method:</th>
        <td>GET</td>
      </tr>
      <tr>
        <th>Request URL:</th>
        <td>http://localhost:8000/breach/get_work/1</td>
      </tr>
      
      <tr>
        <th>Raised by:</th>
        <td>breach.views.get_work</td>
      </tr>
      
    </table>
  </div>
  <div id="info">
    
      <p>Victim not found</p>
    
  </div>

  <div id="explanation">
    <p>
      You're seeing this error because you have <code>DEBUG = True</code> in
      your Django settings file. Change that to <code>False</code>, and Django
      will display a standard 404 page.
    </p>
  </div>
</body>
</html>

```
undefined:2
<!DOCTYPE html>
^

SyntaxError: Unexpected token <
    at Object.parse (native)
    at IncomingMessage.<anonymous> (/Users/test/M3/rupture-develop/realtime/index.js:45:45)
    at emitNone (events.js:72:20)
    at IncomingMessage.emit (events.js:166:7)
    at endReadableNT (_stream_readable.js:913:12)
    at nextTickCallbackWith2Args (node.js:442:9)
    at process._tickCallback (node.js:356:17)

npm ERR! Darwin 14.1.0
npm ERR! argv "/usr/local/bin/node" "/usr/local/bin/npm" "start"
npm ERR! node v4.4.3
npm ERR! npm  v2.15.1
npm ERR! code ELIFECYCLE
npm ERR! rupture@1.0.0 start: `node --harmony --harmony_destructuring index.js`
npm ERR! Exit status 1
npm ERR! 
npm ERR! Failed at the rupture@1.0.0 start script 'node --harmony --harmony_destructuring index.js'.
npm ERR! This is most likely a problem with the rupture package,
npm ERR! not with npm itself.
npm ERR! Tell the author that this fails on your system:
npm ERR!     node --harmony --harmony_destructuring index.js
npm ERR! You can get information on how to open an issue for this project with:
npm ERR!     npm bugs rupture
npm ERR! Or if that isn't available, you can get their info via:
npm ERR! 
npm ERR!     npm owner ls rupture
npm ERR! There is likely additional logging output above.

npm ERR! Please include the following file with any support request:
npm ERR!     /Users/test/M3/rupture-develop/realtime/npm-debug.log

--- a/realtime/index.js
+++ b/realtime/index.js
@@ -18,7 +18,7 @@ const BACKEND_HOST = 'localhost',
 
 socket.on('connection', function(client) {
     winston.info('New connection from client ' + client.id);
-    
+
     var victimId;
     client.on('client-hello', function({victim_id}) {
         if (!victims.victim_id) {
@@ -47,8 +47,12 @@ socket.on('connection', function(client) {
                 responseData += chunk;
             });
             response.on('end', function() {
-                winston.info('Got (get-work) response from backend: ' + responseData);
-                client.emit('do-work', JSON.parse(responseData));
+                try {
+                    client.emit('do-work', JSON.parse(responseData));
+                    winston.info('Got (get-work) response from backend: ' + responseData);
+                } catch (e) {
+                    winston.error('Got invalid (get-work) response from backend');
+                }
             });
         });
         getWorkRequest.on('error', function(err) {

--- a/realtime/index.js
+++ b/realtime/index.js
@@ -83,7 +83,11 @@ socket.on('connection', function(client) {
             });
             response.on('end', function() {
                 winston.info('Got (work-completed) response from backend: ' + responseData);
-                var victory = JSON.parse(responseData)['victory'];
+                try {
+                    var victory = JSON.parse(responseData)['victory'];
+                } catch (e) {
+                    winston.error('Got invalid (work-completed) response from backend');
+                }
                 if (victory === false) {
                     createNewWork();
                 }

